apollo-router/src/plugins/rhai.rs

```diff
@@ -1431,6 +1482,29 @@ impl Rhai {
                 *x = Uri::from_parts(parts).map_err(|e| e.to_string())?;
                 Ok(())
             })
+            // Request.uri.host
+            .register_get_result("host", |x: &mut Uri| to_dynamic(x.host()))
```

Comment:
Probably, but this is the sort of thing I'd like to see accessible from the `http_compat` layer, so that this code would become:
```
http_compat::get_host(uri)
```
and then we could use that from Rust plugins (or anywhere else in the router). However, I don't want to do that work as part of this change (and it probably requires more discussion/design).

Reply:
I'd rather have it done properly right away, otherwise the API would have to change again

Reply:
I'm hesitating because it feels a bit "wrong" just getting the HOST header when the user is trying to interact with the `request.uri`. Maybe I should provide a rhai fn which does this:
```
fn get_request_host(request) {
    if request.uri.host() == () {
        return request.headers["HOST"];
    }
}
```

Reply:
I don't think there is a right or wrong answer here, but I'm trying to decide the best approach by extending principles.

I like the principle of least surprise. It seems odd to me that if a user does an attribute access, we may potentially give them a different value. It's convenient, but it may be surprising and I don't like surprising users. It may be very surprising to someone writing tests or for some odd reason not wanting to take a host from a header.

I think the best compromise is to leave the attribute access as is, but provide a convenience function (as described above, with docs) that explains that it will try the attribute first and then look for a HOST header if not in the URI.

Reply:
This is in the context of a HTTP request where there are precise rules on where we should get the host from and with which priority. We must follow those rules